orm/model_bucket.go

```diff
@@ -0,0 +1,174 @@
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is retured and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest ModelSlicePtr) error
+
+	// Put saves given model in the database. Before inserting into
+	// database, model is validated using its Validate method.
+	// If the key is nil or zero length then a sequence generator is used
+	// to create a unique key value.
+	// Using a key that alreayd exists in the database cause the value to
+	// be overwritten.
+	Put(db weave.KVStore, key []byte, m Model) ([]byte, error)
+
+	// Delete removes an entity with given primary key from the database.
+	// It returns ErrNotFound if an entity with given key does not exist.
+	Delete(db weave.KVStore, key []byte) error
+
+	// Register registers this buckets content to be accessable via query
+	// requests under the given name.
+	Register(name string, r weave.QueryRouter)
+}
+
+// NewModelBucket returns a ModelBucket instance. This implementation relies on
+// a bucket instance. Final implementation should operate directly on the
+// KVStore instead.
+func NewModelBucket(b Bucket) ModelBucket {
+	return &modelBucket{
+		b:     b,
+		idSeq: b.Sequence("id"),
+	}
+}
+
+type modelBucket struct {
+	b     Bucket
+	idSeq Sequence
+}
+
+func (mb *modelBucket) Register(name string, r weave.QueryRouter) {
+	mb.b.Register(name, r)
+}
+
+func (mb *modelBucket) One(db weave.ReadOnlyKVStore, key []byte, dest Model) error {
+	obj, err := mb.b.Get(db, key)
+	if err != nil {
+		return err
+	}
+	if obj == nil || obj.Value() == nil {
+		return errors.Wrapf(errors.ErrNotFound, "%T not in the store", dest)
+	}
+	res := obj.Value()
+
+	if !reflect.TypeOf(res).AssignableTo(reflect.TypeOf(dest)) {
+		return errors.Wrapf(errors.ErrType, "%T cannot be represented as %T", res, dest)
+	}
+
+	reflect.ValueOf(dest).Elem().Set(reflect.ValueOf(res).Elem())
+	return nil
+}
+
+func (mb *modelBucket) ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, destination ModelSlicePtr) error {
+	objs, err := mb.b.GetIndexed(db, indexName, key)
+	if err != nil {
+		return err
+	}
+	if len(objs) == 0 {
+		return nil
+	}
+
+	dest := reflect.ValueOf(destination)
+	if dest.Kind() != reflect.Ptr {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of models")
+	}
+	if dest.IsNil() {
+		return errors.Wrap(errors.ErrImmutable, "got nil pointer")
+	}
+	dest = dest.Elem()
+	if dest.Kind() != reflect.Slice {
+		return errors.Wrap(errors.ErrType, "destination must be a pointer to slice of models")
+	}
+
+	// It is allowed to pass destination as both []MyModel and []*MyModel
+	sliceOfPointers := dest.Type().Elem().Kind() == reflect.Ptr
```

Comment:
:smile: 

Either you had fun playing with reflect, or sweated fighting with it.
But end result is nice

Reply:
:finnadie:

This is not the first time I use `reflect` like this, so it's not that painful anymore. This is not the Go I like to write though.